File: SOUNDBOY/Audio/EqPresets.cs
using System.Text;

namespace SOUNDBOY.Audio;

public sealed record EqPreset(string Name, float[] Bands, float Preamp);

public static class EqPresets
{
    public static readonly EqPreset[] BuiltIn =
    {
        new("Classical",          new[] { 0f, 0, 0, 0, 0, 0, -7.2f, -7.2f, -7.2f, -9.6f }, 0),
        new("Club",               new[] { 0f, 0, 8, 5.6f, 5.6f, 5.6f, 3.2f, 0, 0, 0 }, 0),
        new("Dance",              new[] { 9.6f, 7.2f, 2.4f, 0, 0, -5.6f, -7.2f, -7.2f, 0, 0 }, 0),
        new("Full Bass",          new[] { -8f, 9.6f, 9.6f, 5.6f, 1.6f, -4, -8, -10.4f, -11.2f, -11.2f }, -3),
        new("Full Bass & Treble", new[] { 7.2f, 5.6f, 0, -7.2f, -4.8f, 1.6f, 8, 11.2f, 12, 12 }, -3),
        new("Full Treble",        new[] { -9.6f, -9.6f, -9.6f, -4, 2.4f, 11.2f, 12, 12, 12, 12 }, -3),
        new("Laptop Speakers",    new[] { 4.8f, 11.2f, 5.6f, -3.2f, -2.4f, 1.6f, 4.8f, 9.6f, 12, 12 }, -3),
        new("Large Hall",         new[] { 10.4f, 10.4f, 5.6f, 5.6f, 0, -4.8f, -4.8f, -4.8f, 0, 0 }, -2),
        new("Live",               new[] { -4.8f, 0, 4, 5.6f, 5.6f, 5.6f, 4, 2.4f, 2.4f, 2.4f }, 0),
        new("Party",              new[] { 7.2f, 7.2f, 0, 0, 0, 0, 0, 0, 7.2f, 7.2f }, 0),
        new("Pop",                new[] { -1.6f, 4.8f, 7.2f, 8, 5.6f, 0, -2.4f, -2.4f, -1.6f, -1.6f }, 0),
        new("Reggae",             new[] { 0f, 0, 0, -5.6f, 0, 6.4f, 6.4f, 0, 0, 0 }, 0),
        new("Rock",               new[] { 8f, 4.8f, -5.6f, -8, -3.2f, 4, 8.8f, 11.2f, 11.2f, 11.2f }, -3),
        new("Ska",                new[] { -2.4f, -4.8f, -4, 0, 4, 5.6f, 8.8f, 9.6f, 11.2f, 9.6f }, -2),
        new("Soft",               new[] { 4.8f, 1.6f, 0, -2.4f, 0, 4, 8, 9.6f, 11.2f, 12 }, -2),
        new("Soft Rock",          new[] { 4f, 4, 2.4f, 0, -4, -5.6f, -3.2f, 0, 2.4f, 8.8f }, 0),
        new("Techno",             new[] { 8f, 5.6f, 0, -5.6f, -4.8f, 0, 8, 9.6f, 9.6f, 8.8f }, -2),
    };

    // Winamp .EQF: 31-byte header, then records of 257-byte name + 11 bytes (10 bands, preamp).
    // Each byte is a slider position 0..63 where 0 is the top (+12 dB) and 31 is 0 dB.
    static readonly byte[] Header = Encoding.ASCII.GetBytes("Winamp EQ library file v1.1\u001a!--");
    const int NameLen = 257;

    static float ToDb(byte b) => Math.Clamp((31 - b) / 31f * 12f, -12f, 12f);
    static byte ToByte(float db) => (byte)Math.Clamp((int)Math.Round(31 - db / 12f * 31), 0, 63);

    public static List<EqPreset> ReadEqf(string path)
    {
        var data = File.ReadAllBytes(path);
        if (data.Length < Header.Length || !data.AsSpan(0, 27).SequenceEqual(Header.AsSpan(0, 27)))
            throw new InvalidDataException("Not a Winamp EQ library (.eqf) file.");
        var list = new List<EqPreset>();
        for (int pos = Header.Length; pos + NameLen + 11 <= data.Length; pos += NameLen + 11)
        {
            int end = Array.IndexOf(data, (byte)0, pos, NameLen);
            string name = Encoding.Latin1.GetString(data, pos, (end < 0 ? pos + NameLen : end) - pos).Trim();
            var bands = new float[10];
            for (int i = 0; i < 10; i++) bands[i] = ToDb(data[pos + NameLen + i]);
            list.Add(new EqPreset(name.Length > 0 ? name : "Imported", bands, ToDb(data[pos + NameLen + 10])));
        }
        return list;
    }

    public static void WriteEqf(string path, IEnumerable<EqPreset> presets)
    {
        using var fs = File.Create(path);
        fs.Write(Header);
        foreach (var p in presets)
        {
            var name = new byte[NameLen];
            var raw = Encoding.Latin1.GetBytes(p.Name);
            Array.Copy(raw, name, Math.Min(raw.Length, NameLen - 1));
            fs.Write(name);
            for (int i = 0; i < 10; i++) fs.WriteByte(ToByte(p.Bands[i]));
            fs.WriteByte(ToByte(p.Preamp));
        }
    }
}

File: mac/Sources/SoundboyCore/EqPresets.swift
import Foundation

public enum Equalizer {
    public static let bandCount = 10
    public static let maxDb: Float = 12
    public static let frequencies: [Float] = [60, 170, 310, 600, 1000, 3000, 6000, 12000, 14000, 16000]
    public static let labels = ["60", "170", "310", "600", "1K", "3K", "6K", "12K", "14K", "16K"]
}

public struct EqPreset: Equatable {
    public var name: String
    public var bands: [Float]
    public var preamp: Float
    public init(name: String, bands: [Float], preamp: Float) { self.name = name; self.bands = bands; self.preamp = preamp }
}

public enum EqPresets {
    public static let builtIn: [EqPreset] = [
        .init(name: "Classical", bands: [0, 0, 0, 0, 0, 0, -7.2, -7.2, -7.2, -9.6], preamp: 0),
        .init(name: "Club", bands: [0, 0, 8, 5.6, 5.6, 5.6, 3.2, 0, 0, 0], preamp: 0),
        .init(name: "Dance", bands: [9.6, 7.2, 2.4, 0, 0, -5.6, -7.2, -7.2, 0, 0], preamp: 0),
        .init(name: "Full Bass", bands: [-8, 9.6, 9.6, 5.6, 1.6, -4, -8, -10.4, -11.2, -11.2], preamp: -3),
        .init(name: "Full Bass & Treble", bands: [7.2, 5.6, 0, -7.2, -4.8, 1.6, 8, 11.2, 12, 12], preamp: -3),
        .init(name: "Full Treble", bands: [-9.6, -9.6, -9.6, -4, 2.4, 11.2, 12, 12, 12, 12], preamp: -3),
        .init(name: "Laptop Speakers", bands: [4.8, 11.2, 5.6, -3.2, -2.4, 1.6, 4.8, 9.6, 12, 12], preamp: -3),
        .init(name: "Large Hall", bands: [10.4, 10.4, 5.6, 5.6, 0, -4.8, -4.8, -4.8, 0, 0], preamp: -2),
        .init(name: "Live", bands: [-4.8, 0, 4, 5.6, 5.6, 5.6, 4, 2.4, 2.4, 2.4], preamp: 0),
        .init(name: "Party", bands: [7.2, 7.2, 0, 0, 0, 0, 0, 0, 7.2, 7.2], preamp: 0),
        .init(name: "Pop", bands: [-1.6, 4.8, 7.2, 8, 5.6, 0, -2.4, -2.4, -1.6, -1.6], preamp: 0),
        .init(name: "Reggae", bands: [0, 0, 0, -5.6, 0, 6.4, 6.4, 0, 0, 0], preamp: 0),
        .init(name: "Rock", bands: [8, 4.8, -5.6, -8, -3.2, 4, 8.8, 11.2, 11.2, 11.2], preamp: -3),
        .init(name: "Ska", bands: [-2.4, -4.8, -4, 0, 4, 5.6, 8.8, 9.6, 11.2, 9.6], preamp: -2),
        .init(name: "Soft", bands: [4.8, 1.6, 0, -2.4, 0, 4, 8, 9.6, 11.2, 12], preamp: -2),
        .init(name: "Soft Rock", bands: [4, 4, 2.4, 0, -4, -5.6, -3.2, 0, 2.4, 8.8], preamp: 0),
        .init(name: "Techno", bands: [8, 5.6, 0, -5.6, -4.8, 0, 8, 9.6, 9.6, 8.8], preamp: -2),
    ]

    // Winamp .EQF: 31-byte header, then records of 257-byte name + 11 bytes (10 bands, preamp).
    // Each byte is a slider position 0..63 where 0 is the top (+12 dB) and 31 is 0 dB.
    static let header = Array("Winamp EQ library file v1.1\u{1A}!--".utf8)
    static let nameLength = 257

    static func toDb(_ b: UInt8) -> Float { min(12, max(-12, (31 - Float(b)) / 31 * 12)) }
    static func toByte(_ db: Float) -> UInt8 { UInt8(min(63, max(0, (31 - db / 12 * 31).rounded()))) }

    public static func readEqf(_ url: URL) throws -> [EqPreset] {
        let data = [UInt8](try Data(contentsOf: url))
        guard data.count >= header.count, Array(data[0..<27]) == Array(header[0..<27]) else {
            throw NSError(domain: "SOUNDBOY", code: 1, userInfo: [NSLocalizedDescriptionKey: "Not a Winamp EQ library (.eqf) file."])
        }
        var list: [EqPreset] = []
        var pos = header.count
        while pos + nameLength + 11 <= data.count {
            let nameBytes = data[pos..<(pos + nameLength)].prefix { $0 != 0 }
            let name = String(decoding: nameBytes, as: UTF8.self).trimmingCharacters(in: .whitespaces)
            let bands = (0..<10).map { toDb(data[pos + nameLength + $0]) }
            list.append(EqPreset(name: name.isEmpty ? "Imported" : name, bands: bands, preamp: toDb(data[pos + nameLength + 10])))
            pos += nameLength + 11
        }
        return list
    }

    public static func writeEqf(_ presets: [EqPreset], to url: URL) throws {
        var out = header
        for p in presets {
            var name = [UInt8](repeating: 0, count: nameLength)
            let raw = Array(p.name.utf8.prefix(nameLength - 1))
            name.replaceSubrange(0..<raw.count, with: raw)
            out += name
            out += (0..<10).map { toByte(p.bands[$0]) }
            out.append(toByte(p.preamp))
        }
        try Data(out).write(to: url)
    }
}
